collectors/profile.plugin/plugin_profile.cc

```diff
@@ -0,0 +1,165 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+#include "plugin_profile.h"
+
+#include <chrono>
+#include <memory>
+#include <vector>
+#include <random>
+#include <sstream>
+#include <string>
+#include <map>
+
+class Dimension {
+public:
+    Dimension(RRDDIM *RD, size_t Index) :
+        RD(RD), Index(Index), NumUpdates(0) {}
+
+    void update(const struct timeval &TV, size_t Iteration) {
+        UNUSED(Iteration);
+
+        collected_number CN = Index + NumUpdates++;
+        rrddim_timed_set_by_pointer(RD->rrdset, RD, TV, CN);
+    }
+
+private:
+    RRDDIM *RD;
+    size_t Index;
+    size_t NumUpdates;
+};
+
+class Chart {
+    static size_t ChartIdx;
+
+public:
+    static std::shared_ptr<Chart> create(const char *Id, size_t UpdateEvery) {
+        RRDSET *RS = rrdset_create(
+            localhost,
+            "prof_type",
+            Id, // id
+            NULL, // name
+            "prof_family",
+            NULL, // ctx
+            Id, // title
+            "prof_units",
+            "prof_plugin",
+            "prof_module",
+            41000 + ChartIdx++,
+            static_cast<int>(UpdateEvery),
+            RRDSET_TYPE_LINE
+        );
+
+        if (!RS)
+            fatal("Could not create chart %s", Id);
+
+        return std::make_shared<Chart>(RS);
+    }
+
+public:
+    Chart(RRDSET *RS) : RS(RS), Initialized(false) {}
+
+    void createDimension(const char *Name, size_t Index) {
+        RRDDIM *RD = rrddim_add(
+            RS, Name, NULL, 1, 1, RRD_ALGORITHM_ABSOLUTE
+        );
+
+        if (!RD)
+            fatal("Could not create dimension %s.%s", rrdset_id(RS), Name);
+
+        Dimensions.push_back(std::make_shared<Dimension>(RD, Index));
+    }
+
+    void update(const struct timeval &TV, size_t Iteration) {
+        if (Initialized) {
+            usec_t DurationSinceLastUpdate = RS->update_every * USEC_PER_SEC;
+            rrdset_timed_next(RS, TV, DurationSinceLastUpdate);
+        }
+
+        Initialized = true;
+
+        for (std::shared_ptr<Dimension> &D : Dimensions)
+            D->update(TV, Iteration);
+
+        rrdset_timed_done(RS, TV, /* pending_rrdset_next */ false);
+    }
+
+private:
+    RRDSET *RS;
+    bool Initialized;
+    std::vector<std::shared_ptr<Dimension>> Dimensions;
+};
+
+size_t Chart::ChartIdx = 0;
+
+static std::vector<std::shared_ptr<Chart>> createCharts(size_t NumCharts, unsigned NumDimsPerChart, time_t UpdateEvery) {
+    std::vector<std::shared_ptr<Chart>> Charts;
+    Charts.reserve(NumCharts);
+
+    constexpr size_t Len = 1024;
+    char Buf[Len];
+
+    for (size_t ChartIdx = 0; ChartIdx != NumCharts; ChartIdx++) {
+        snprintfz(Buf, Len, "chart_%zu", ChartIdx + 1);
```

Comment:
In normal situation when we call this function, instead to give as argument the buffer size, we give a smaller size as you can see [here](https://github.com/netdata/netdata/blob/76d93161cd041774e95a23eac00057437a7f07e6/collectors/apps.plugin/apps_plugin.c#L1136-L1137).

Of course with the given size, I do not expect issues. I am only calling attention for our pattern.